src/vos/vos_iterator.c

```diff
@@ -749,50 +819,42 @@ vos_iterate_internal(vos_iter_param_t *param, vos_iter_type_t type,
 
 			reset_anchors(iter_ent.ie_child_type, anchors);
 
-			/** The child iterator may yield during iteration.
-			 *  It isn't necessary to reprobe the parent in order
-			 *  to continue iterating on the child because it
-			 *  would remain in the same location in memory.
-			 *  However, it is necessary to reprobe this iterator
-			 *  before continuing.
-			 */
-			if (need_reprobe(type, anchors)) {
-				D_ASSERT(!daos_anchor_is_zero(anchor) &&
-					 !daos_anchor_is_eof(anchor));
-				goto probe;
-			}
+			rc = advance_stage(type, 0, param, anchors, anchor, &stage,
+					   VOS_ITER_STAGE_POST, &probe_flags);
+			JUMP_TO_STAGE(rc, next, probe, out);
```

Comment:
It's back from subtree iteration and in parent tree now, and if we need re-probe in parent tree, we'll do re-probe with 'VOS_ITER_PROBE_AGAIN', and the whole subtree will be iterated again, right?

Reply:
No, the stage will get set to VOS_ITER_STAGE_POST if we are going to re-probe so it will skip the RECURSE stage.

Reply:
oh, I see. That explains why you have both 'stage check' and 'goto label'.